### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by alsrgv**
* Update benchmarks.
* NCCL now supports RoCE.
* Update NCCL installation instructions.
* Add new troubleshooting instructions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -215,12 +234,13 @@ command.
 
 1. Install [NCCL 2](https://developer.nvidia.com/nccl).
 
-If you aren't able to install NCCL 2 Debian package due to missing dependencies, you can use this workaround:
+Steps to install NCCL 2 are listed [here](http://docs.nvidia.com/deeplearning/sdk/nccl-install-guide/index.html).
+
+If you have installed NCCL 2 using `nccl-<version>.txz` package, you should add library path to `LD_LIBRARY_PATH`
+environment variable or register it in `/etc/ld.so.conf`.
```
Comment (jeevandev) [inline README.md]: Nit: Can we have NCCL installation steps at one place only, so that we can avoid duplicate & does not have to worry about steps going out of sync? Similar comment on other duplicated instructions for installation

File: README.md [hunk 2]
```diff
@@ -22,25 +22,35 @@ averaging gradients across them, `tf.Server()`, `tf.ClusterSpec()`, `tf.train.Sy
...
 |               On 16 GPUs              |                   x |                   x |                   x |
-| Distributed TensorFlow                |     1,378.4 (10.4x) |        996.8 (8.4x) |        310.4 (2.4x) |
-| Distributed TensorFlow (vars. on CPU) |     1,586.0 (11.9x) |     1,195.2 (10.1x) |        299.2 (2.3x) |
-| TCP Horovod on CPU                    |     2,003.2 (15.1x) |     1,232.0 (10.4x) |        696.0 (5.3x) |
-| RDMA Horovod on CPU                   | **2,068.8 (15.6x)** |     1,590.4 (13.5x) |        752.0 (5.7x) |
-| TCP Horovod on GPU (NCCL)             |     1,921.6 (14.4x) |     1,475.2 (12.5x) |     1,635.2 (12.5x) |
-| RDMA Horovod on GPU (NCCL)            |     1,974.4 (14.8x) | **1,651.2 (14.0x)** | **1,824.0 (13.9x)** |
+| Distributed TensorFlow                |     1,345.8 (10.0x) |        959.6 (8.0x) |         74.7 (0.6x) |
+| Distributed TensorFlow (vars. on CPU) |     1,576.4 (11.7x) |      1,168.8 (9.8x) |         79.5 (0.6x) |
+| TCP Horovod on CPU                    | **2,073.3 (15.4x)** |     1,338.3 (11.2x) |        616.8 (4.7x) |
```
Comment (jeevandev) [inline README.md]: Do you think clarifying what does "* Horovod on CPU" mean would be better?

File: README.md [hunk 3]
```diff
@@ -395,6 +424,59 @@ Or:
 $ HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_NCCL_INCLUDE=/path/to/nccl/include HOROVOD_NCCL_LIB=/path/to/nccl/lib pip install --no-cache-dir horovod
 ```
 
+### NCCL 2 is not found during runtime
```
Comment (jeevandev) [inline README.md]: Not for this review but eventually we might want to move all this to a separate docs folder & have it linked like how it is done in https://github.com/uber/pyflame (check FAQ link).

File: README.md [hunk 4]
```diff
@@ -22,25 +22,35 @@ averaging gradients across them, `tf.Server()`, `tf.ClusterSpec()`, `tf.train.Sy
...
+| Baseline single-GPU (batch size=64)   |               134.4 |               119.4 |               130.9 |
 |               On 16 GPUs              |                   x |                   x |                   x |
-| Distributed TensorFlow                |     1,378.4 (10.4x) |        996.8 (8.4x) |        310.4 (2.4x) |
-| Distributed TensorFlow (vars. on CPU) |     1,586.0 (11.9x) |     1,195.2 (10.1x) |        299.2 (2.3x) |
-| TCP Horovod on CPU                    |     2,003.2 (15.1x) |     1,232.0 (10.4x) |        696.0 (5.3x) |
-| RDMA Horovod on CPU                   | **2,068.8 (15.6x)** |     1,590.4 (13.5x) |        752.0 (5.7x) |
-| TCP Horovod on GPU (NCCL)             |     1,921.6 (14.4x) |     1,475.2 (12.5x) |     1,635.2 (12.5x) |
-| RDMA Horovod on GPU (NCCL)            |     1,974.4 (14.8x) | **1,651.2 (14.0x)** | **1,824.0 (13.9x)** |
+| Distributed TensorFlow                |     1,345.8 (10.0x) |        959.6 (8.0x) |         74.7 (0.6x) |
+| Distributed TensorFlow (vars. on CPU) |     1,576.4 (11.7x) |      1,168.8 (9.8x) |         79.5 (0.6x) |
```
Comment (sblotner) [inline README.md]: What's vars. ? Variables? Will people know what this means?

File: README.md [hunk 5]
```diff
@@ -22,25 +22,35 @@ averaging gradients across them, `tf.Server()`, `tf.ClusterSpec()`, `tf.train.Sy
...
+| Setup                                 |     Inception V3    |      ResNet-101     |        VGG-16       |
+|---------------------------------------|:-------------------:|:-------------------:|:-------------------:|
+| Baseline single-GPU (batch size=64)   |               148.8 |               136.0 |               149.6 |
+|               On 64 GPUs              |                   x |                   x |                   x |
+| Distributed TensorFlow                |     4,225.3 (28.4x) |     2,996.0 (22.0x) |         97.0 (0.6x) |
+| Distributed TensorFlow (vars. on CPU) |     5,297.4 (35.6x) |     4,269.2 (31.4x) |        100.8 (0.7x) |
+| TCP Horovod on CPU                    |     6,549.6 (44.0x) |     3,761.6 (27.7x) |      1,462.6 (9.8x) |
+| TCP Horovod on GPU (NCCL)             | **7,932.1 (53.3x)** | **7,741.6 (56.9x)** | **6,084.2 (40.7x)** |
+
+While installing MPI and NCCL itself may seem like an extra hassle, it only needs to be done once and by team dealing
```
Comment (sblotner) [inline README.md]: once and by team --> once by the team

File: README.md [hunk 6]
```diff
@@ -22,25 +22,35 @@ averaging gradients across them, `tf.Server()`, `tf.ClusterSpec()`, `tf.train.Sy
...
+with infrastructure, while everyone else in the company who builds the models can enjoy simplicity of training them at
```
Comment (sblotner) [inline README.md]: the simplicity

File: README.md [hunk 7]
```diff
@@ -172,16 +182,17 @@ To use Horovod on GPU, read the options below and see which one applies to you b
...
 In most situations, using NCCL 2 will significantly improve performance over the CPU version.  NCCL 2 provides the *allreduce*
-operation optimized for NVIDIA GPUs and a variety of networking devices, such as InfiniBand.
+operation optimized for NVIDIA GPUs and a variety of networking devices, such as RoCE or InfiniBand.
 
 1. Install [NCCL 2](https://developer.nvidia.com/nccl).
 
-If you aren't able to install NCCL 2 Debian package due to missing dependencies, you can use this workaround:
+Steps to install NCCL 2 are listed [here](http://docs.nvidia.com/deeplearning/sdk/nccl-install-guide/index.html).
+
+If you have installed NCCL 2 using `nccl-<version>.txz` package, you should add library path to `LD_LIBRARY_PATH`
```
Comment (sblotner) [inline README.md]: 2 using --> 2 using the 
add library path to -- >add the library path to the

File: README.md [hunk 8]
```diff
@@ -190,13 +201,21 @@ Steps to install Open MPI are listed [here](https://www.open-mpi.org/faq/?catego
 
 3. Install the `horovod` pip package.
 
+If you have installed NCCL 2 using `nccl-<version>.txz` package, you should specify path to NCCL 2 using the `HOROVOD_NCCL_HOME`
```
Comment (sblotner) [inline README.md]: 2 using --> 2 using the 
specify path --> specify the path

File: README.md [hunk 9]
```diff
@@ -190,13 +201,21 @@ Steps to install Open MPI are listed [here](https://www.open-mpi.org/faq/?catego
...
+environment variable.
+
+```bash
+$ HOROVOD_NCCL_HOME=/usr/local/nccl-<version> HOROVOD_GPU_ALLREDUCE=NCCL pip install --no-cache-dir horovod
+```
+
+If you have installed NCCL 2 using Ubuntu package, you can simply run:
```
Comment (sblotner) [inline README.md]: 2 using --> 2 using the

File: README.md [hunk 10]
```diff
@@ -215,12 +234,13 @@ command.
 
 1. Install [NCCL 2](https://developer.nvidia.com/nccl).
 
-If you aren't able to install NCCL 2 Debian package due to missing dependencies, you can use this workaround:
+Steps to install NCCL 2 are listed [here](http://docs.nvidia.com/deeplearning/sdk/nccl-install-guide/index.html).
+
+If you have installed NCCL 2 using `nccl-<version>.txz` package, you should add library path to `LD_LIBRARY_PATH`
```
Comment (sblotner) [inline README.md]: 2 using --> 2 using the 
add library --> add the library

File: README.md [hunk 11]
```diff
@@ -234,6 +254,15 @@ sure you build it with [CUDA support](https://www.open-mpi.org/faq/?category=bui
 
 4. Install the `horovod` pip package.
 
+If you have installed NCCL 2 using `nccl-<version>.txz` package, you should specify path to NCCL 2 using the `HOROVOD_NCCL_HOME`
```
Comment (sblotner) [inline README.md]: 2 using --> 2 using the
specify path --> specify the path

File: README.md [hunk 12]
```diff
@@ -234,6 +254,15 @@ sure you build it with [CUDA support](https://www.open-mpi.org/faq/?category=bui
...
+environment variable.
+
+```bash
+$ HOROVOD_NCCL_HOME=/usr/local/nccl-<version> HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=MPI HOROVOD_GPU_BROADCAST=MPI pip install --no-cache-dir horovod
+```
+
+If you have installed NCCL 2 using Ubuntu package, you can simply run:
```
Comment (sblotner) [inline README.md]: using Ubuntu --> using the Ubuntu

File: README.md [hunk 13]
```diff
@@ -395,6 +424,59 @@ Or:
 $ HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_NCCL_INCLUDE=/path/to/nccl/include HOROVOD_NCCL_LIB=/path/to/nccl/lib pip install --no-cache-dir horovod
 ```
 
+### NCCL 2 is not found during runtime
+
+If you see the error message below, it means NCCL 2 was not found in standard libraries location. You should add directory
```
Comment (sblotner) [inline README.md]: in standard --> in the standard
should add --> should add the

File: README.md [hunk 14]
```diff
@@ -395,6 +424,59 @@ Or:
...
+where you installed NCCL 2 libraries to `LD_LIBRARY_PATH` environment variable.
```
Comment (sblotner) [inline README.md]: libraries to -- >libraries to the

File: README.md [hunk 15]
```diff
@@ -433,8 +515,8 @@ memory, use:
 
 ## Tensor Fusion
 
-One of the unique things about Horovod is its ability to interleave communication and computation with batching
-of small *allreduce* operations, which results in improved performance. We call this batching feature Tensor Fusion.
+One of the unique things about Horovod is its ability to interleave communication and computation coupled with ability
```
Comment (sblotner) [inline README.md]: with the ability

File: README.md (hunk 1 above)
Comment (alsrgv) [reply README.md]: Agreed, seems that separate docs folder would solve this and the other comment below.
